parlai/core/worlds.py

```diff
@@ -975,6 +975,202 @@ def shutdown(self):
+        self.world = world
+        # TODO: maybe generalize this
+        self.max_size = self.truncate * opt['batchsize']
+
+        # buffer worlds
+        self.worlds = [
+            shared['world_class'](opt, None, shared) for _ in range(self._BUFFER_SIZE)
+        ]
+
+        self.reset()
+
+    def reset(self):
+        super().reset()
+        self._obs = [None for _ in range(self._BUFFER_SIZE)]
+        self._scores = [None for _ in range(self._BUFFER_SIZE)]
+
+        self.number_parleys = 0
+        self.total_exs = 0
+        self.world.reset()
+        for w in self.worlds:
+            w.reset()
+
+    def reset_metrics(self):
+        super().reset_metrics()
+        self.world.reset_metrics()
+        for w in self.worlds:
+            w.reset_metrics()
+
+    def epoch_done(self):
+        return all(w.epoch_done() for w in self.worlds)
+
+    def num_examples(self):
+        return self.world.num_examples()
+
+    def num_episodes(self):
+        return self.world.num_episodes()
+
+    def _ceil(self, n, r=4):
+        """
+        Round to the nearest multiple of 4.
+
+        TensorCores only work when a tensor is a multiple of 4 in all dimensions. This
+        means all examples cost is related to their nearest multiple of 4.
+        """
+        # round up to 4, all things are equal
+        return ((n + r - 1) // r) * r
+
+    def _score(self, obs):
+        if 'text_vec' in obs:
+            # note that all examples have a cost that is based on their
+            # nearest multiple of 4. We can therefore mix-and-match
+            # anything with the same cost for increased stochasticity,
+            # while not really wasting much padding.
+            return self._ceil(len(obs['text_vec']))
+        else:
+            return None
+
+    def parley(self):
+        # first make sure that all the worlds are processed in the queue
+        assert not self.epoch_done()
+        indices = []
+        for i in range(self._BUFFER_SIZE):
+            if self._scores[i] is not None:
+                indices.append(i)
+                continue
+            if self.worlds[i].epoch_done():
+                continue
+
+            act = self.worlds[i].agents[0].act()
+            obs = self.worlds[i].agents[1].observe(act)
+            self._obs[i] = obs
+
+            if act is None:
+                # TODO: we might need to do something special here around the end of
+                # the epoch
+                raise TypeError("This is unexpected.")
+
+            self._scores[i] = self._score(obs)
+            if self._scores[i] is not None:
+                indices.append(i)
+
+        assert len(indices) != 0, "DynamicBatchWorld ran out of data!"
+
+        for i in indices:
+            assert self._scores is not None
+
+        # sort all the indices by their score, so that we can find similarly lengthed
+        # items in O(1)
+        indices = sorted(indices, key=lambda i: (self._scores[i], random.random()))
```

Comment:
what is the random for? just for shuffling in the case of a tie?

Reply:
Yes, maximize stochasticity for SGD reasons